identity-iota/src/credential/presentation_validator.rs

```diff
@@ -0,0 +1,849 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Borrow;
+use std::collections::BTreeMap;
+use std::marker::PhantomData;
+
+use identity_core::common::OneOrMany;
+use identity_credential::credential::Credential;
+use identity_credential::presentation::Presentation;
+use identity_did::verifiable::VerifierOptions;
+use serde::Serialize;
+
+use super::errors::AccumulatedCredentialValidationError;
+use super::errors::AccumulatedPresentationValidationError;
+use super::errors::ValidationError;
+use super::PresentationValidationOptions;
+use crate::credential::credential_validator::CredentialValidator;
+use crate::did::IotaDID;
+use crate::document::ResolvedIotaDocument;
+use crate::Error;
+use crate::Result;
+
+/// A struct for validating the given [Presentation].
+pub struct PresentationValidator<U, V, T: Borrow<Presentation<U, V>>> {
+  pub presentation: T,
+  _required: (PhantomData<U>, PhantomData<V>), // will not compile without this field
+}
```

Comment:
I think both `CredentialValidator` and `PresentationValidator` could be replaced with empty structs and have all validation logic left as static methods that take `&Credential`/`Borrow<Credential<T>>` and `&Presentation`/`Borrow<Presentation>` directly since the validators hold no state or options themselves.

This would make exposing them to the Wasm bindings simple as we would not have to deal with borrowed structs and instead just expose static associated methods.

Reply:
Basically the advantage would be calling the functions independently for cases where users don't want to go through full validation, e.g.

```rust
CredentialValidator.earliest_expiry_date(&credential, Timestamp::now_utc())?;
```

instead of 

```rust
let validator = CredentialValidator::new(&credential);
validator.earliest_expiry_date(Timestamp::now_utc())?;
```

Reply:
What you are suggesting is more in the spirit of "making easy things easy and hard things possible" which is a good thing. As much as I dislike using structs for grouping related things alone (preferring to use the module system for this), I think it is fine in this case as we may want to introduce methods in the future that require some pre-computed state.

Reply:
> As much as I dislike using structs for grouping related things alone (preferring to use the module system for this)

Yeah I'm of the complete opposite opinion and consider free-floating functions API pollution. I even want to make a struct to group the encoding methods like `encode_b58`, `encode_multibase` etc.

> we may want to introduce methods in the future that require some pre-computed state.

Good point. If that's necessary in the future most checks could still stay as static associated methods to be used independently.